File: app/src/main/java/dev/ragnarok/fenrir/place/Place.java
package dev.ragnarok.fenrir.place;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.activity.result.ActivityResultLauncher;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentResultListener;

import dev.ragnarok.fenrir.util.Objects;

public class Place implements Parcelable {

    public static final int VIDEO_PREVIEW = 1;
    public static final int FRIENDS_AND_FOLLOWERS = 2;
    public static final int EXTERNAL_LINK = 3;
    public static final int DOC_PREVIEW = 4;
    public static final int WALL_POST = 5;
    public static final int COMMENTS = 6;
    public static final int WALL = 7;
    public static final int CONVERSATION_ATTACHMENTS = 8;
    public static final int PLAYER = 9;
    public static final int SEARCH = 10;
    public static final int CHAT = 11;
    public static final int BUILD_NEW_POST = 12;
    public static final int EDIT_COMMENT = 13;
    public static final int EDIT_POST = 14;
    public static final int REPOST = 15;
    public static final int DIALOGS = 16;
    public static final int FORWARD_MESSAGES = 17;
    public static final int TOPICS = 18;
    public static final int CHAT_MEMBERS = 19;
    public static final int COMMUNITIES = 20;
    public static final int LIKES_AND_COPIES = 21;
    public static final int VIDEO_ALBUM = 22;
    public static final int AUDIOS = 23;
    public static final int VIDEOS = 24;
    public static final int VK_PHOTO_ALBUMS = 25;
    public static final int VK_PHOTO_ALBUM = 26;
    public static final int VK_PHOTO_ALBUM_GALLERY = 27;
    public static final int VK_PHOTO_ALBUM_GALLERY_SAVED = 28;

    public static final int FAVE_PHOTOS_GALLERY = 29;
    public static final int SIMPLE_PHOTO_GALLERY = 30;
    public static final int POLL = 31;
    public static final int PREFERENCES = 32;
    public static final int DOCS = 33;
    public static final int FEED = 34;
    public static final int NOTIFICATIONS = 35;
    public static final int BOOKMARKS = 36;
    public static final int RESOLVE_DOMAIN = 37;
    public static final int VK_INTERNAL_PLAYER = 38;
    public static final int NOTIFICATION_SETTINGS = 39;
    public static final int CREATE_PHOTO_ALBUM = 40;
    public static final int EDIT_PHOTO_ALBUM = 41;
    public static final int MESSAGE_LOOKUP = 42;
    public static final int GIF_PAGER = 43;
    public static final int SECURITY = 44;
    public static final int CREATE_POLL = 45;
    public static final int COMMENT_CREATE = 46;
    public static final int LOGS = 47;
    public static final int LOCAL_IMAGE_ALBUM = 48;
    public static final int SINGLE_SEARCH = 49;
    public static final int NEWSFEED_COMMENTS = 50;
    public static final int COMMUNITY_CONTROL = 51;
    public static final int COMMUNITY_BAN_EDIT = 52;
    public static final int COMMUNITY_ADD_BAN = 53;

    public static final int VK_PHOTO_TMP_SOURCE = 54;

    public static final int COMMUNITY_MANAGER_EDIT = 55;
    public static final int COMMUNITY_MANAGER_ADD = 56;

    public static final int REQUEST_EXECUTOR = 57;
    public static final int USER_BLACKLIST = 58;

    public static final int PROXY_ADD = 59;
    public static final int DRAWER_EDIT = 60;
    public static final int SIDE_DRAWER_EDIT = 61;
    public static final int USER_DETAILS = 62;
    public static final int AUDIOS_IN_ALBUM = 63;
    public static final int COMMUNITY_INFO = 64;
    public static final int COMMUNITY_INFO_LINKS = 65;
    public static final int SETTINGS_THEME = 66;
    public static final int SEARCH_BY_AUDIO = 67;
    public static final int MENTIONS = 68;
    public static final int OWNER_ARTICLES = 69;
    public static final int WALL_ATTACHMENTS = 70;
    public static final int STORY_PLAYER = 71;
    public static final int SINGLE_PHOTO = 72;
    public static final int ARTIST = 73;
    public static final int CATALOG_BLOCK_AUDIOS = 74;
    public static final int CATALOG_BLOCK_PLAYLISTS = 75;
    public static final int CATALOG_BLOCK_VIDEOS = 76;
    public static final int CATALOG_BLOCK_LINKS = 77;
    public static final int SHORT_LINKS = 78;
    public static final int IMPORTANT_MESSAGES = 79;
    public static final int MARKET_ALBUMS = 80;
    public static final int MARKETS = 81;
    public static final int MARKET_VIEW = 82;
    public static final int GIFTS = 83;
    public static final int PHOTO_ALL_COMMENT = 84;
    public static final int ALBUMS_BY_VIDEO = 85;
    public static final int FRIENDS_BY_PHONES = 86;
    public static final int UNREAD_MESSAGES = 87;
    public static final int AUDIOS_SEARCH_TABS = 88;
    public static final int GROUP_CHATS = 89;
    public static final int LOCAL_SERVER_PHOTO = 90;
    public static final int VK_PHOTO_ALBUM_GALLERY_NATIVE = 91;
    public static final int SEARCH_COMMENTS = 92;
    public static final Creator<Place> CREATOR = new Creator<Place>() {
        @Override
        public Place createFromParcel(Parcel in) {
            return new Place(in);
        }

        @Override
        public Place[] newArray(int size) {
            return new Place[size];
        }
    };
    private final int type;
    private boolean needFinishMain;
    private String requestListenerKey;
    private FragmentResultListener requestListener;
    private ActivityResultLauncher<Intent> activityResultLauncher;
    private Bundle args;

    public Place(int type) {
        this.type = type;
    }

    protected Place(Parcel in) {
        type = in.readInt();
        args = in.readBundle(getClass().getClassLoader());
    }

    public void tryOpenWith(@NonNull Context context) {
        if (context instanceof PlaceProvider) {
            ((PlaceProvider) context).openPlace(this);
        }
    }

    public Place setFragmentListener(@NonNull String requestListenerKey, @NonNull FragmentResultListener requestListener) {
        this.requestListenerKey = requestListenerKey;
        this.requestListener = requestListener;
        return this;
    }

    public Place setActivityResultLauncher(@NonNull ActivityResultLauncher<Intent> activityResultLauncher) {
        this.activityResultLauncher = activityResultLauncher;
        return this;
    }

    public boolean isNeedFinishMain() {
        return needFinishMain;
    }

    public Place setNeedFinishMain(boolean needFinishMain) {
        this.needFinishMain = needFinishMain;
        return this;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(type);
        dest.writeBundle(args);
    }

    public Bundle getArgs() {
        return args;
    }

    public Place setArguments(Bundle arguments) {
        args = arguments;
        return this;
    }

    public Place withStringExtra(String name, String value) {
        prepareArguments().putString(name, value);
        return this;
    }

    public Place withParcelableExtra(String name, Parcelable parcelableExtra) {
        prepareArguments().putParcelable(name, parcelableExtra);
        return this;
    }

    public Place withIntExtra(String name, int value) {
        prepareArguments().putInt(name, value);
        return this;
    }

    public Place withLongExtra(String name, long value) {
        prepareArguments().putLong(name, value);
        return this;
    }

    public Bundle prepareArguments() {
        if (args == null) {
            args = new Bundle();
        }

        return args;
    }

    public void applyFragmentListener(@NonNull Fragment fragment, @NonNull FragmentManager fragmentManager) {
        if (Objects.nonNull(requestListener)) {
            fragmentManager.setFragmentResultListener(requestListenerKey, fragment, requestListener);
        }
    }

    public void launchActivityForResult(@NonNull Activity context, @NonNull Intent intent) {
        if (Objects.nonNull(activityResultLauncher) && !needFinishMain) {
            activityResultLauncher.launch(intent);
        } else {
            context.startActivity(intent);
            if (needFinishMain) {
                context.finish();
                context.overridePendingTransition(0, 0);
            }
        }
    }

    public int getType() {
        return type;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/place/Place.kt
package dev.ragnarok.fenrir.place

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.activity.result.ActivityResultLauncher
import dev.ragnarok.fenrir.util.Utils

open class Place : Parcelable {
    val type: Int
    var isNeedFinishMain = false
        private set
    private var activityResultLauncher: ActivityResultLauncher<Intent>? = null
    private var args: Bundle? = null

    constructor(type: Int) {
        this.type = type
    }

    protected constructor(p: Parcel) {
        type = p.readInt()
        args = p.readBundle(javaClass.classLoader)
    }

    fun tryOpenWith(context: Context) {
        if (context is PlaceProvider) {
            (context as PlaceProvider).openPlace(this)
        }
    }

    fun setActivityResultLauncher(activityResultLauncher: ActivityResultLauncher<Intent>): Place {
        this.activityResultLauncher = activityResultLauncher
        return this
    }

    fun setNeedFinishMain(needFinishMain: Boolean): Place {
        isNeedFinishMain = needFinishMain
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(type)
        dest.writeBundle(args)
    }

    fun setArguments(arguments: Bundle?): Place {
        args = arguments
        return this
    }

    fun withStringExtra(name: String, value: String?): Place {
        prepareArguments().putString(name, value)
        return this
    }

    fun withParcelableExtra(name: String, parcelableExtra: Parcelable?): Place {
        prepareArguments().putParcelable(name, parcelableExtra)
        return this
    }

    fun withIntExtra(name: String, value: Int): Place {
        prepareArguments().putInt(name, value)
        return this
    }

    fun withBoolExtra(name: String, value: Boolean): Place {
        prepareArguments().putBoolean(name, value)
        return this
    }

    fun withLongExtra(name: String, value: Long): Place {
        prepareArguments().putLong(name, value)
        return this
    }

    fun prepareArguments(): Bundle {
        if (args == null) {
            args = Bundle()
        }
        return args!!
    }

    fun safeArguments(): Bundle {
        return args ?: Bundle()
    }

    fun launchActivityForResult(context: Activity, intent: Intent) {
        if (activityResultLauncher != null && !isNeedFinishMain) {
            activityResultLauncher?.launch(intent)
        } else {
            context.startActivity(intent)
            if (isNeedFinishMain) {
                Utils.finishActivityImmediate(context)
            }
        }
    }

    companion object {
        const val VIDEO_PREVIEW = 1
        const val FRIENDS_AND_FOLLOWERS = 2
        const val EXTERNAL_LINK = 3
        const val DOC_PREVIEW = 4
        const val WALL_POST = 5
        const val COMMENTS = 6
        const val WALL = 7
        const val CONVERSATION_ATTACHMENTS = 8
        const val PLAYER = 9
        const val SEARCH = 10
        const val CHAT = 11
        const val BUILD_NEW_POST = 12
        const val EDIT_COMMENT = 13
        const val EDIT_POST = 14
        const val REPOST = 15
        const val DIALOGS = 16
        const val FORWARD_MESSAGES = 17
        const val TOPICS = 18
        const val CHAT_MEMBERS = 19
        const val COMMUNITIES = 20
        const val LIKES_AND_COPIES = 21
        const val VIDEO_ALBUM = 22
        const val AUDIOS = 23
        const val VIDEOS = 24
        const val VK_PHOTO_ALBUMS = 25
        const val VK_PHOTO_ALBUM = 26
        const val VK_PHOTO_ALBUM_GALLERY_NATIVE = 27
        const val VK_PHOTO_ALBUM_GALLERY = 28
        const val VK_PHOTO_ALBUM_GALLERY_SAVED = 29
        const val VK_PHOTO_TMP_SOURCE = 30
        const val FAVE_PHOTOS_GALLERY = 31
        const val SIMPLE_PHOTO_GALLERY = 32
        const val SINGLE_PHOTO = 33
        const val POLL = 34
        const val PREFERENCES = 35
        const val DOCS = 36
        const val FEED = 37
        const val NOTIFICATIONS = 38
        const val BOOKMARKS = 39
        const val RESOLVE_DOMAIN = 40
        const val VK_INTERNAL_PLAYER = 41
        const val NOTIFICATION_SETTINGS = 42
        const val CREATE_PHOTO_ALBUM = 43
        const val EDIT_PHOTO_ALBUM = 44
        const val MESSAGE_LOOKUP = 45
        const val GIF_PAGER = 46
        const val SECURITY = 47
        const val CREATE_POLL = 48
        const val COMMENT_CREATE = 49
        const val LOGS = 50
        const val LOCAL_IMAGE_ALBUM = 51
        const val SINGLE_SEARCH = 52
        const val NEWSFEED_COMMENTS = 53
        const val COMMUNITY_CONTROL = 54
        const val COMMUNITY_BAN_EDIT = 55
        const val COMMUNITY_ADD_BAN = 56
        const val COMMUNITY_MANAGER_EDIT = 57
        const val COMMUNITY_MANAGER_ADD = 58
        const val REQUEST_EXECUTOR = 59
        const val USER_BLACKLIST = 60
        const val PROXY_ADD = 61
        const val DRAWER_EDIT = 62
        const val SIDE_DRAWER_EDIT = 63
        const val USER_DETAILS = 64
        const val AUDIOS_IN_ALBUM = 65
        const val COMMUNITY_INFO = 66
        const val COMMUNITY_INFO_LINKS = 67
        const val SETTINGS_THEME = 68
        const val SEARCH_BY_AUDIO = 69
        const val MENTIONS = 70
        const val OWNER_ARTICLES = 71
        const val WALL_ATTACHMENTS = 72
        const val STORY_PLAYER = 73
        const val ARTIST = 74
        const val SHORT_LINKS = 75
        const val IMPORTANT_MESSAGES = 76
        const val MARKET_ALBUMS = 77
        const val MARKETS = 78
        const val MARKET_VIEW = 79
        const val GIFTS = 80
        const val PHOTO_ALL_COMMENT = 81
        const val ALBUMS_BY_VIDEO = 82
        const val FRIENDS_BY_PHONES = 83
        const val UNREAD_MESSAGES = 84
        const val AUDIOS_SEARCH_TABS = 85
        const val GROUP_CHATS = 86
        const val LOCAL_SERVER_PHOTO = 87
        const val SEARCH_COMMENTS = 88
        const val SHORTCUTS = 89
        const val NARRATIVES = 90
        const val VOTERS = 91
        const val FEED_BAN = 92
        const val REMOTE_FILE_MANAGER = 93
        const val COMMUNITY_MEMBERS = 94
        const val FRIENDS_BIRTHDAYS = 95
        const val CATALOG_V2_AUDIO_CATALOG = 96
        const val CATALOG_V2_AUDIO_SECTION = 97
        const val CATALOG_V2_LIST_EDIT = 98
        const val STORIES_VIEWS = 99
        const val SHORT_VIDEOS = 100

        @JvmField
        val CREATOR: Parcelable.Creator<Place> = object : Parcelable.Creator<Place> {
            override fun createFromParcel(p: Parcel): Place {
                return Place(p)
            }

            override fun newArray(size: Int): Array<Place?> {
                return arrayOfNulls(size)
            }
        }
    }
}
